### DISCUSSION TITLE
AboutBDT: I can't recv a large datagram with datagram

### DISCUSSION THREAD

**Discussion by streetycat** [buckyos/CYFS#105] [Q&A]
I build a network protocol with bdt.datagram in my project:

I start 4 nodes in one process, One CyfsStack is opened in every node with 127.0.0.1, and they are broadcast packages to each others.

But the package is discarded frequently, That is irrational in one process.

The source code is below:

1. create tunnel
```
let datagram = bdt_stack.datagram_manager().bind(NET_PROTOCOL_VPORT)?;
```
3. send package
```
        let mut options = DatagramOptions::default();

        self.datagram
            .send_to(buf.as_slice(), &mut options, &remote, self.vport);
```
5. recv package
```
            match datagram.recv_v().await {
                Ok(pkgs) => {
                    for datagram in pkgs {
                        let remote = datagram.source.remote.object_id().clone();
                        match HotstuffPackage::raw_decode(datagram.data.as_slice()) {
                            Ok((pkg, remain)) => {
                                log::debug!(
                                    "[group-listener] {:?}-{} recv message from {:?}, msg: {:?}, len: {}",
                                    pkg.rpath(),
                                    local_device_id,
                                    remote,
                                    pkg,
                                    datagram.data.len()
                                );
                                assert_eq!(remain.len(), 0);
                                processor.on_message(pkg, remote).await;
                            }
                            Err(err) => {
                                log::debug!(
                                    "[group-listener] {} recv message from {:?}, len: {} decode failed {:?}",
                                    local_device_id,
                                    remote,
                                    datagram.data.len(),
                                    err
                                );
                            }
                        }
                    }
                }
                Err(e) => log::warn!("group listener failed: {:?}", e),
            }
```

**Comment by streetycat** [comment]
I found that that will happen more frequently for large package over 1KB.

**Reply by jing-git** [reply]
When the data packet is larger than the MTU, it will be fragmented during transmission. You can set a unique sequence value to reassemble the complete packet after receiving the fragments:
let mut options = DatagramOptions::default();
options.sequence = Some(TempSeq::from(cyfs_base::bucky_time_now() as u32));
self.datagram
            .send_to(buf.as_slice(), &mut options, &remote, self.vport);

**Reply by streetycat** [reply]
> When the data packet is larger than the MTU, it will be fragmented during transmission. You can set a unique sequence value to reassemble the complete packet after receiving the fragments: let mut options = DatagramOptions::default(); options.sequence = Some(TempSeq::from(cyfs_base::bucky_time_now() as u32)); self.datagram .send_to(buf.as_slice(), &mut options, &remote, self.vport);

Thank you, I'm finished it follow your suggestion.

But I have another questions for it.

What's the value of MTU?
Did you mean that, the reassemble process will depend on the parameter(options.sequence)? I think it will be more friendly to support it inner.
